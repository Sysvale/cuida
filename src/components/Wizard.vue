<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<div class="cds-wizard">
		<div class="cds-wizard__container">
			<div
				v-for="(step, index) in steps"
				:key="`wizard-step-${index}`"
				class="cds-wizard__box"
				:class="{ 'cds-wizard__empty-state-box': currentStep < index, 'cds-wizard__elevated': currentStep === index }"
				fluid
				:clickable="clickable"
				@box-click="handleStepBoxClick(index)"
				@click.prevent
			>
				<span v-if="currentStep >= index">
					<slot :name="`step-${index + 1}-header`">
						<h3 class="cds-wizard__title">
							{{ step.title }}
						</h3>
						<h4 class="cds-wizard__subtitle">
							{{ step.subtitle }}
						</h4>
					</slot>
					<div class="cds-wizard__content">
						<slot :name="`step-${index + 1}`" />
					</div>
				</span>
				<div
					v-else
					class="cds-wizard__empty-state"
				>
					<img
						v-if="step.image"
						class="cds-wizard__empty-state-image"
						:src="step.image"
						:alt="`Imagem de ${step.title}`"
					>
					<div
						class="cds-wizard__empty-state-title"
					>
						{{ step.title }}
					</div>
					<div
						class="cds-wizard__empty-state-subtitle"
					>
						{{ emptyStateText(step.title) }}
					</div>
				</div>
			</div>
		</div>
		<div class="cds-wizard__buttons">
			<cds-button
				:text="cancelButtonText"
				secondary
				@button-click="backActionClick"
				@click.prevent
			/>
			<cds-button
				:text="nextButtonText"
				:variant="nextButtonVariant"
				:disabled="disableNextButton"
				@button-click="nextActionClick"
				@click.prevent
			/>
		</div>
	</div>
</template>

<script>
import CdsBox from './Box.vue';
import CdsButton from './Button.vue';

export default {
	components: {
		CdsBox,
		CdsButton,
	},

	props: {
		/**
		 * Um array com objetos com as propriedades 'title' (obrigatório), 'subtitle' e 'image'.
		 * O 'title' e 'subtitle' são textos que descrevem a etapa, a 'imagem' é o caminho para
		 * a imagem do empty-state;
		 */
		steps: {
			type: Array,
			default: () => ([]),
			required: true,
			validator: (value) => value.length >= 2 && value.length <= 3,
		},
		/**
		 * O índice da etapa atual (0, 1 ou 2)
		 */
		activeStep: {
			type: Number,
			default: 0,
			validator: (value) => (value >= 0 && value <= 2),
		},
		/**
		* Ativa ou desativa o clique no passo (etapa) para ir para esta etapa
		*/
		clickable: {
			type: Boolean,
			default: false,
		},
		/**
		 * O texto do botão de avançar (ação principal).
		 */
		nextButtonText: {
			type: String,
			default: 'Próximo',
		},
		/**
		 * O texto do botão de voltar (ação secundária).
		 */
		cancelButtonText: {
			type: String,
			default: 'Anterior',
		},
		/**
		 * A variante do botão de ação principal. São 10 variantes: 'teal', 'green', 'blue',
		 * 'indigo', 'violet', 'pink', 'red', 'orange', 'amber' e 'dark'.
		 */
		nextButtonVariant: {
			type: String,
			default: 'green',
		},
		/**
		 * Controla a disponibilidade do botão de ação principal.
		 */
		disableNextButton: {
			type: Boolean,
			default: false,
		},
	},

	data() {
		return {
			currentStep: this.activeStep,
		};
	},

	computed: {
		disabledBack() {
			return this.currentStep <= 0;
		},

		disabledNext() {
			return this.currentStep >= this.steps.length - 1;
		},
	},

	watch: {
		activeStep: {
			handler(newValue, oldValue) {
				if (newValue !== oldValue) {
					this.currentStep = newValue;
				}
			},
			immediate: true,
		},

		currentStep(newValue, oldValue) {
			if (newValue !== oldValue) {
				/**
				 * Evento emitido quando a etapa ativa é alterada
				* @event step-change
				* @type {Event}
				*/
				this.$emit('step-change', newValue);
			}
		},
	},

	methods: {
		emptyStateText(title) {
			return `Preencha todos os campos obrigatórios da etapa anterior para
				liberar a seção de ${title.toLowerCase()}`;
		},

		handleStepBoxClick(index) {
			this.currentStep = index;
		},

		backActionClick() {
			if (this.currentStep > 0) {
				this.currentStep--;
			} else {
				/**
				* Evento que indica que a ação de cancelar foi acionada.
				* Emitido ao clicar no botão de voltar na primeira etapa.
				* @event cancel-action
				* @type {Event}
				*/
				this.$emit('cancel-action', true);
			}
		},

		nextActionClick() {
			let nextStep = this.currentStep;
			if (this.currentStep < this.steps.length - 1) {
				nextStep++;
			}
			/**
			* Evento que indica que ação de avançar foi acionada
			* @event next-action
			* @type {Event}
			*/
			this.$emit('next-action', { nextStep });
		},
	},
};
</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

::v-deep .cds-wizard__empty-state-box .box__container {
	height: 100% !important;
}
::v-deep .cds-wizard__empty-state-box .spacer {
	display: flex !important;
	height: 100% !important;
}

.cds-wizard {
	&__container {
		display: flex;
		gap: spacer(3);
	}

	&__box {
		background-color: $n-0;
		border: 1px solid $n-30;
		border-radius: $border-radius-medium;
		padding: pa(5);
		height: inherit;
		width: 100%;
	}

	&__elevated {
		box-shadow: $shadow-sm;
	}

	&__empty-state {
		display: flex;
		flex-direction: column;
		text-align: center;
		justify-content: center;
		align-items: center;
		margin: mYX(0, 10);

		&-title {
			@include subheading-3;
			color: $n-600;
			margin: mb(2);
		}

		&-subtitle {
			@include caption;
			color: $n-400;
		}

		&-image {
			width: 75px;
			height: auto;
			margin: mb(5);
		}
	}

	&__buttons {
		display: flex;
		justify-content: flex-end;
		margin: mt(5);
		gap: spacer(3);
	}

	&__title {
		@include subheading-2;
		color: $n-700;
		margin: mb(1);
	}

	&__subtitle {
		@include body-2;
		color: $n-400;
		margin: ma(0);
	}

	&__content {
		margin: mYX(6, 0);
	}
}

@media (max-width: 992px) {
	.cds-wizard {
		&__container {
			flex-direction: column;
		}
	}
}
</style>
