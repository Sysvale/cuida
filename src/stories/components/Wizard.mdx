import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Wizard from '../../components/Wizard.vue';
import TextInput from '../../components/TextInput.vue';
import * as WizardStories from './Wizard.stories';

<Meta of={WizardStories} />

export const Template = (args) => ({
	components: { CdsWizard: Wizard, CdsTextInput: TextInput },
	setup() {
		return { args };
	},
	data() {
		return {
			activeStep: 0,
		};
	},
	template: /*html*/
		`<cds-wizard
			v-bind="args"
			:active-step="activeStep"
			@step-change="handleStepChange"
			@next-action="handleNextAction"
			@cancel-action="handleCancelAction"
		>
			<template #step-1>
				<div style="height: 400px; gap: 16px; display: flex; flex-direction: column;">
					<cds-text-input
						fluid
						label="Nome"
					/>
					<cds-text-input
						fluid
						label="Sobrenome"
					/>
					<cds-text-input
						fluid
						label="Nome da mãe"
					/>
				</div>
			</template>
			<template #step-2>
				<div style="height: 400px; gap: 16px; display: flex; flex-direction: column;">
					<cds-text-input
						fluid
						label="CEP"
					/>
					<cds-text-input
						fluid
						label="Rua"
					/>
					<cds-text-input
						fluid
						label="Bairro"
					/>
					<cds-text-input
						fluid
						label="Cidade"
					/>
				</div>
			</template>
			<template #step-3>
				<div style="height: 400px; gap: 16px; display: flex; flex-direction: column;">
					<cds-text-input
						fluid
						label="Modalidade"
					/>
					<cds-text-input
						fluid
						label="URL do site"
					/>
					<cds-text-input
						fluid
						label="Informação Complementar"
					/>
				</div>
			</template>
		</cds-wizard>
		<br>
		`,
	methods: {
		handleStepChange(value) {
			console.info('%c@step-change: ', 'color: #2C70CD;', value);
			this.activeStep = value;
		},
		handleNextAction(value) {
			console.info('%c@next-action: ', 'color: #2C70CD;', value.nextStep);
			this.activeStep = value.nextStep;
		},
		handleCancelAction(value) {
			console.info('%c@cancel-action: ', 'color: #2C70CD;', value);
		}
	},
});

# Wizard

### Wizards são componentes utilizados para um processo passo a passo que permite ao usuário inserir informações em uma ordem prescrita e na qual as etapas subsequentes podem depender das informações inseridas nas etapas anteriores.

***

<br />

## Quando usar:

* Quando precisar completar etapas em uma determinada sequência.
* For necessário exibir informações em  blocos para simplificar o fluxo.

<br />

## Quando não usar:

* Quando o processo não for tão complexo, não havendo necessidade de divisão em etapas.
* Quando houver mais de três etapas, deve se considerar dividir em mais processos.
* Quando não houver necessidade de guiar o usuário numa sequência.

<br />

## Preview

export const steps = [
	{
		title: 'Informações gerais',
		subtitle: 'Insira as informações de identificação',
	},
	{
		title: 'Endereço de entrega',
	},
	{
		title: 'Informações complementares',
		subtitle: 'Adicione informações complementares para triagem',
		image: 'https://static.vecteezy.com/system/resources/previews/011/537/753/non_2x/box-empty-state-single-isolated-icon-with-flat-style-free-vector.jpg',
	},
];

<Canvas>
  <Story of={WizardStories.Wizard} />
</Canvas>

<Controls />
