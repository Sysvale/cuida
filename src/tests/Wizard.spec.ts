import { describe, test, expect } from 'vitest';
import Wizard from '../components/Wizard.vue';
import { mount } from '@vue/test-utils';

const mockedSteps = [
	{
		title: 'Dummy title 1',
		subtitle: 'Dummy subtitle 1',
		completed: true,
		inProcessing: false,
		error: false,
	},
	{
		title: 'Dummy title 2',
		subtitle: 'Dummy subtitle 2',
		completed: false,
		inProcessing: false,
		error: true,
	},
	{
		title: 'Dummy title 3',
		subtitle: 'Dummy subtitle 3',
		completed: false,
		inProcessing: true,
		error: false,
	},
];

describe('Wizard', () => {
	test('renders correctly', async () => {
		const wrapper = mount(Wizard, {
			props: {
				steps: mockedSteps,
			},
			slots: {
				'step-1-header': 'Dummy title without subtitle',
				'step-1': '<p>Dummy content 1</p>',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('it accepts valid steps length', () => {
		const validator = Wizard.props.steps.validator;
		const invalidSteps = [...mockedSteps, {}];
		expect(validator(mockedSteps)).toBe(true);
		expect(validator(invalidSteps)).toBe(false);
	});
});
