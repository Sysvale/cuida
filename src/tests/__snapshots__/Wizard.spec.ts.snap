// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wizard > renders correctly 1`] = `
"<div data-v-61a7952a="" class="cds-wizard">
  <div data-v-61a7952a="" class="cds-wizard__container">
    <div data-v-61a7952a="" class="cds-wizard__box cds-wizard__elevated" fluid="" clickable="false"><span data-v-61a7952a="">Dummy title without subtitle<div data-v-61a7952a="" class="cds-wizard__content"><p>Dummy content 1</p></div></span></div>
    <div data-v-61a7952a="" class="cds-wizard__box cds-wizard__empty-state-box" fluid="" clickable="false">
      <div data-v-61a7952a="" class="cds-wizard__empty-state">
        <!--v-if-->
        <div data-v-61a7952a="" class="cds-wizard__empty-state-title">Dummy title 2</div>
        <div data-v-61a7952a="" class="cds-wizard__empty-state-subtitle">Preencha todos os campos obrigatórios da etapa anterior para
          liberar a seção de dummy title 2</div>
      </div>
    </div>
    <div data-v-61a7952a="" class="cds-wizard__box cds-wizard__empty-state-box" fluid="" clickable="false">
      <div data-v-61a7952a="" class="cds-wizard__empty-state">
        <!--v-if-->
        <div data-v-61a7952a="" class="cds-wizard__empty-state-title">Dummy title 3</div>
        <div data-v-61a7952a="" class="cds-wizard__empty-state-subtitle">Preencha todos os campos obrigatórios da etapa anterior para
          liberar a seção de dummy title 3</div>
      </div>
    </div>
  </div>
  <div data-v-61a7952a="" class="cds-wizard__buttons"><button data-v-3c9d0845="" data-v-61a7952a="" class="button__container button--secondary--active button-size--md">
      <!--v-if-->
      <!-- @slot Slot padrão utilizado para exibir texto do botão. -->Anterior
    </button><button data-v-3c9d0845="" data-v-61a7952a="" class="button__container button--green--active button-size--md">
      <!--v-if-->
      <!-- @slot Slot padrão utilizado para exibir texto do botão. -->Próximo
    </button></div>
</div>"
`;
